src/Common/src/Interop/Windows/winhttp/Interop.SafeWinHttpHandle.cs

```diff
@@ -30,6 +30,31 @@ public static void DisposeAndClearHandle(ref SafeWinHttpHandle winHttpHandler)
                 }
             }
 
+            // The SafeHandle base class is inconsistent w.r.t. how Dispose() and DangerousRelease()
+            // are implemented. Both methods can cause the ref count on the SafeHandle to be decremented.
+            // And when the ref count reaches zero, ReleaseHandle() is called and the handle is marked
+            // as 'closed'. However, there is a separate internal state bit to track Dispose() calls.
+            // Calling Dispose() more than once is a no-op.  However, calling DangerousRelease() and then
+            // Dispose() will cause ObjectDisposedException to be thrown if the previous DangerousRelease()
+            // caused the handle to be closed. This seems inconsistent especially since both methods can
+            // decrement the ref count and cause ReleaseHandle() to be called.
```

Comment:
@stephentoub As my tests show, both Dispose() and DangerousRelease() will cause the ref counts to decrement and both will cause the handle to be freed when the count reaches 0.

We really need a more purer way of doing AddRef/Release that is always balanced similar to COM.  That model allows handles to be shared among other objects and for them to be freed when all references go to zero.  The current SafeHandle base class is not very flexible for those requirements.

Do you have other suggestions for "sharing" SafeHandles among the related WinHttpHandler classes?